framework/src/kernels/MaterialDerivativeTestKernel.C

```diff
@@ -0,0 +1,66 @@
+/****************************************************************/
+/*               DO NOT MODIFY THIS HEADER                      */
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*           (c) 2010 Battelle Energy Alliance, LLC             */
+/*                   ALL RIGHTS RESERVED                        */
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+
+#include "MaterialDerivativeTestKernel.h"
+
+
+template<>
+InputParameters validParams<MaterialDerivativeTestKernel>()
+{
+  InputParameters params = validParams<Kernel>();
+  params.addClassDescription("Class used for testing derivatives of a material property.");
+  params.addRequiredParam<MaterialPropertyName>("material_property", "Name of material property for which derivatives are to be tested.");
+  params.addRequiredCoupledVar("vars", "List of nonlinear variables.");
+
+  return params;
+}
+
+MaterialDerivativeTestKernel::MaterialDerivativeTestKernel(const InputParameters & parameters) :
+    DerivativeMaterialInterface<Kernel>(parameters),
+    _p_name(getParam<MaterialPropertyName>("material_property")),
+    _p(getMaterialPropertyByName<Real>(_p_name))
+{
+  _n_vars = coupledComponents("vars");
+
+  _p_derivatives.resize(_n_vars);
+  _var_nums.resize(_n_vars);
+  for (unsigned int m = 0; m < _n_vars; ++m)
+  {
+    _var_nums[m] = coupled("vars", m);
+    MooseVariable * var = getVar("vars", m);
+    _p_derivatives[m] = & getMaterialPropertyDerivativeByName<Real>(_p_name, var->name());
+  }
+}
+
+Real
+MaterialDerivativeTestKernel::computeQpResidual()
+{
+  return _p[_qp] * _test[_i][_qp];
+}
+
+Real
+MaterialDerivativeTestKernel::computeQpJacobian()
+{
+  return computeQpOffDiagJacobian(_var.number());
```

Comment:
I think Daniel was originally suggesting that I use a new nonlinear variable for this kernel to act upon for each test, so I guess that's why he's suggesting returning zero. What I've done is to use an arbitrary existing nonlinear variable - I don't know which approach is preferred.